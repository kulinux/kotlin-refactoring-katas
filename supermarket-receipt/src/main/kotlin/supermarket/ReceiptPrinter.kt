package supermarket

import supermarket.model.Discount
import supermarket.model.Receipt
import supermarket.model.ReceiptItem
import java.util.*

private const val CARRIAGE_RETURN = "\n"

class ReceiptPrinter @JvmOverloads constructor(private val columns: Int = 40) {
  fun printReceipt(receipt: Receipt): String {
    return handleItems(receipt.getItems()) +
      handleDiscounts(receipt.getDiscounts()) +
      handleTotal(receipt.totalPrice)
  }

  private fun handleItems(items: List<ReceiptItem>): String {
    return items.joinToString("") { handleItem(it) }
  }

  private fun handleDiscounts(discounts: List<Discount>): String {
    return discounts.joinToString("") { handleDiscount(it) }
  }

  private fun handleTotal(totalPrice: Double): String {
    val pricePresentation = String.format(Locale.UK, "%.2f", totalPrice)
    val total = "Total: "
    val whitespaces = whiteSpacesOfLength(this.columns - total.length - pricePresentation.length)

    return "$CARRIAGE_RETURN$total$whitespaces$pricePresentation"
  }


  private fun handleDiscount(discount: Discount): String {
    val productPresentation = discount.products.joinToString { it.name }
    val pricePresentation = String.format(Locale.UK, "%.2f", discount.discountAmount)
    val description = discount.description

    val whiteSpacesOfLength =
      whiteSpacesOfLength(this.columns - 3 - productPresentation.length - description.length - pricePresentation.length)

    return "$description($productPresentation)$whiteSpacesOfLength-$pricePresentation$CARRIAGE_RETURN"
  }


  private fun handleItem(item: ReceiptItem): String {
    val price = String.format(Locale.UK, "%.2f", item.totalPrice)
    val quantity = presentQuantity(item)
    val name = item.product.name
    val unitPrice = String.format(Locale.UK, "%.2f", item.price)

    val whitespaceSize = this.columns - name.length - price.length
    val baseLine = name + whiteSpacesOfLength(whitespaceSize) + price + "$CARRIAGE_RETURN"

    if (item.quantity != 1.0) {
      return "$baseLine  $unitPrice * $quantity$CARRIAGE_RETURN"
    }
    return baseLine
  }

  private fun presentQuantity(item: ReceiptItem): String {
    return if (item.product.isQuantityInt())
      String.format("%x", item.quantity.toInt())
    else
      String.format(Locale.UK, "%.3f", item.quantity)
  }

  private fun whiteSpacesOfLength(whitespaceSize: Int): String {
    val whitespace = StringBuilder()
    for (i in 0 until whitespaceSize) {
      whitespace.append(" ")
    }
    return whitespace.toString()
  }
}
