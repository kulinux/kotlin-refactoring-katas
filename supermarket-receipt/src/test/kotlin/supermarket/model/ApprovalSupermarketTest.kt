package supermarket.model

import org.approvaltests.Approvals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource
import supermarket.ReceiptPrinter

class ApprovalSupermarketTest {
  lateinit var catalog: SupermarketCatalog
  lateinit var toothbrush: Product
  lateinit var apples: Product

  @BeforeEach
  fun setUp() {
    toothbrush = Product.each("toothbrush")
    apples = Product.kilo("apples")
    catalog = catalog(apples to 1.99, toothbrush to 0.99)
  }

  @Test
  fun testPriceEach() {
    val cart = shoppingCart(toothbrush to 2.0)

    val printReceipt = printReceipt(Teller(catalog), cart)
    Approvals.verify(printReceipt)
  }

  @Test
  fun testPriceKilo() {
    val cart = shoppingCart(apples to 2.5)

    val printReceipt = printReceipt(Teller(catalog), cart)
    Approvals.verify(printReceipt)
  }

  @Test
  fun testAddingTwice() {
    val cart = shoppingCart(apples to 2.5, apples to 1.0)

    val printReceipt = printReceipt(Teller(catalog), cart)
    Approvals.verify(printReceipt)
  }

  @ParameterizedTest(name = "Special Offer {0}")
  @EnumSource(SpecialOfferType::class)
  fun testOffers(offerType: SpecialOfferType) {
    val cart = shoppingCart(toothbrush to 6.0, apples to 6.5)

    val teller = Teller(catalog)
    teller.addSpecialOffer(offerType, toothbrush, 10.0)

    val printReceipt = printReceipt(teller, cart)
    Approvals.verify(printReceipt, Approvals.NAMES.withParameters(offerType.name))
  }


  private fun shoppingCart(vararg productsWithQty: Pair<Product, Double>): ShoppingCart =
    productsWithQty.fold(ShoppingCart()) { acc, element -> acc.addItemQuantity(element.first, element.second); acc }

  private fun catalog(vararg productsWithPrice: Pair<Product, Double>): SupermarketCatalog =
    productsWithPrice.fold(FakeCatalog()) { acc, element -> acc.addProduct(element.first, element.second); acc }

  private fun printReceipt(teller: Teller, cart: ShoppingCart): String {
    val receipt = teller.checksOutArticlesFrom(cart)
    val receiptPrinter = ReceiptPrinter()
    val printReceipt = receiptPrinter.printReceipt(receipt)
    return printReceipt
  }

}
