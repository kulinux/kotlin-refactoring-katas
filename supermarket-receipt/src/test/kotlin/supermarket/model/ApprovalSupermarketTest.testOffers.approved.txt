toothbrush                          1.98
  0.99 * 2
apples                              4.98
  1.99 * 2.500
10.0% off(toothbrush)              -0.20

Total:                              6.76